crates/wasm-mutate/src/mutators/peephole/eggsy/mod.rs

```diff
@@ -0,0 +1,691 @@
+            }
+            Lang::I32ShrU(_) => {
+                newfunc.instruction(Instruction::I32ShrU);
+            }
+            Lang::I32Popcnt(_) => {
+                newfunc.instruction(Instruction::I32Popcnt);
+            }
+            Lang::ILoad(_) => {
+                // Do nothing
+                let operators = &operators[insertion_point..=insertion_point + 1/* take to the next operator to save the offset */];
+                let range = (operators[0].1, operators[1].1);
+
+                // Copy the mem operation
+                let raw_data = &info.get_code_section().data[range.0..range.1];
+                newfunc.raw(raw_data.iter().copied());
+            }
+            Lang::Rand => {
+                newfunc.instruction(Instruction::I32Const(rnd.gen()));
+            }
+            Lang::Unfold(_) => {
+                let child = operands[current][0];
+                let node = id_to_node[usize::from(child)];
+
+                match node {
+                    Lang::I32Const(value) => {
+                        let r: i32 = rnd.gen();
+
+                        newfunc.instruction(Instruction::I32Const(r));
+                        newfunc.instruction(Instruction::I32Const((Wrapping(r) - Wrapping(*value)).0));
+                        newfunc.instruction(Instruction::I32Add);
+                    },
+                    _ => panic!("The oeprand for this operator should be a constant, check if the rewriting rule is defined with such conditions")
+                }
+            }
+            Lang::Symbol(s) => {
+                // Copy the byte stream to aavoid mapping
+                let entryidx = symbolsmap[&s.to_string()];
+                let entry = &minidfg.entries[entryidx];
+
+                // Write the symbol in the correct place of the functions
+
+                let bytes = &info.get_code_section().data
+                    [entry.byte_stream_range.start..entry.byte_stream_range.end];
+                newfunc.raw(bytes.iter().copied());
+                //newfunc.instruction(Instruction::I32Const(*v));
+            }
+            _ => {
+                return Err(crate::Error::UnsupportedType(EitherType::Operator(
+                    format!("Eterm {:?} is not implemented", eterm),
+                )))
+            }
+        }
+        Ok(())
+    }
+
+    fn etermtree2waasm(
+        info: &ModuleInfo,
+        rnd: &mut rand::prelude::SmallRng,
+        insertion_point: usize,
+        id_to_node: &Vec<&Lang>,
+        operands: &Vec<Vec<Id>>,
+        current: usize,
+        newfunc: &mut Function,
+        symbolmap: &HashMap<String, usize>,
+        minidfg: &MiniDFG,
+        operators: &Vec<TupleType>,
+    ) -> crate::Result<()> {
+        let root = id_to_node[current];
+
+        // This is a patch, this logic should be here
+        // If root is Unfol...bypass
+
+        if let Lang::Unfold(_) = root {
+            log::debug!("Unfolding a constant yeih !")
+        } else {
```

Comment:
We ensure that the operand of the Unfold is always a constant number. The current implementation writes first the operands and then the root, if we do the same with the unfold root  it means that the constant will be written and thats something we want to avoid. We take the root instead, knowing that its x operand is a constant (otherwise is an error) and then we write (y + z = x)

Reply:
But I agree this pattern should be like this. It should be smoother